File: android/app/src/main/kotlin/com/goalmaker/app/application/update/UpdateCheckResult.kt
package com.goalmaker.app.application.update

import com.goalmaker.app.domain.update.ReleaseArtifact
import com.goalmaker.app.domain.update.ReleaseManifest

/** What "Check for updates" found. */
sealed interface UpdateCheckResult {
    /** No trusted key is built in, so the channel can't be verified (docs/setup). */
    data object NotConfigured : UpdateCheckResult

    /** Development builds never update themselves; install a release build instead. */
    data object DevelopmentBuild : UpdateCheckResult

    data class UpToDate(val latest: String) : UpdateCheckResult

    data class Available(val manifest: ReleaseManifest, val artifact: ReleaseArtifact) : UpdateCheckResult

    data object Untrusted : UpdateCheckResult

    data class Failed(val detail: String) : UpdateCheckResult
}

File: windows/src/GoalMaker.Core/Updates/UpdateCheckResult.cs
namespace GoalMaker.Core.Updates;

/// <summary>What "Check for updates" found.</summary>
public abstract record UpdateCheckResult
{
    private UpdateCheckResult()
    {
    }

    /// <summary>No trusted key is built in, so the channel can't be verified (docs/setup).</summary>
    public sealed record NotConfigured : UpdateCheckResult;

    /// <summary>Development builds never update themselves; install a release build instead.</summary>
    public sealed record DevelopmentBuild : UpdateCheckResult;

    public sealed record UpToDate(string Latest) : UpdateCheckResult;

    public sealed record Available(ReleaseManifest Manifest, ReleaseArtifact Artifact) : UpdateCheckResult;

    public sealed record Untrusted : UpdateCheckResult;

    public sealed record Failed(string Detail) : UpdateCheckResult;
}
